File: src/dotnet/Harness.Engine/AtomicIO.cs
namespace Harness.Engine;

/// <summary>
/// Atomic "final" write for the stores under <c>.harness</c>: writes to a temp file in
/// the SAME directory as the destination and swaps it in via
/// <see cref="File.Move(string, string, bool)"/> with <c>overwrite: true</c> — atomic on
/// the same partition since .NET Core 3.0+. Prevents a crash/kill mid-write from leaving
/// the final file truncated or partially overwritten; a concurrent reader always sees
/// either the complete previous version or the complete new one, never an intermediate
/// state. Doesn't apply to the log/trace <c>File.AppendAllText</c> calls — those are
/// already atomic at the event level (one line, one call) and don't need a file swap.
/// </summary>
internal static class AtomicIO
{
    public static void WriteAllTextAtomic(string path, string content)
    {
        var tmp = TempPathFor(path);
        try
        {
            File.WriteAllText(tmp, content);
            File.Move(tmp, path, overwrite: true);
        }
        catch
        {
            CleanupBestEffort(tmp);
            throw;
        }
    }

    /// <summary>Same atomic guarantee as <see cref="WriteAllTextAtomic"/>, but copying from an existing source file (e.g. snapshotting a live store into its frozen one).</summary>
    public static void CopyAtomic(string sourcePath, string destinationPath)
    {
        var tmp = TempPathFor(destinationPath);
        try
        {
            File.Copy(sourcePath, tmp, overwrite: true);
            File.Move(tmp, destinationPath, overwrite: true);
        }
        catch
        {
            CleanupBestEffort(tmp);
            throw;
        }
    }

    // Unique name per write in the SAME directory as the destination — Path.GetTempFileName()
    // won't do because it creates outside that folder, breaking the atomic-rename guarantee
    // (same partition).
    private static string TempPathFor(string destination) => $"{destination}.tmp-{Guid.NewGuid():N}";

    private static void CleanupBestEffort(string tmp)
    {
        try
        {
            if (File.Exists(tmp))
                File.Delete(tmp);
        }
        catch
        {
            // Cleanup is best-effort — doesn't mask the original exception already being rethrown.
        }
    }
}

File: src/go/harnessengine/atomic_io.go
package harnessengine

import (
	"fmt"
	"os"
	"time"
)

// writeAtomic writes content to path atomically: writes to a unique temp file in the SAME
// directory as the destination, then swaps it in via os.Rename — atomic on the same
// filesystem. A crash/kill mid-write can never leave the final file truncated or partially
// overwritten; a concurrent reader always sees either the previous complete version or the
// new complete one, never an intermediate state. Does not apply to the append-only writes
// of log/trace files — those are already atomic at the event level (one line, one call)
// and don't need a file swap.
func writeAtomic(path, content string) error {
	tmp := tempPathFor(path)
	if err := os.WriteFile(tmp, []byte(content), 0o644); err != nil {
		cleanupBestEffort(tmp)
		return err
	}
	if err := os.Rename(tmp, path); err != nil {
		cleanupBestEffort(tmp)
		return err
	}
	return nil
}

// copyAtomic offers the same atomicity guarantee as writeAtomic, but copies from an
// existing source file (e.g. snapshotting a live store into its frozen counterpart).
func copyAtomic(sourcePath, destinationPath string) error {
	content, err := os.ReadFile(sourcePath)
	if err != nil {
		return err
	}
	return writeAtomic(destinationPath, string(content))
}

// tempPathFor builds a unique name in the SAME directory as destination — os.CreateTemp
// with "" would pick the OS temp dir, breaking the same-filesystem rename guarantee.
func tempPathFor(destination string) string {
	return fmt.Sprintf("%s.tmp-%d-%d", destination, os.Getpid(), time.Now().UnixNano())
}

func cleanupBestEffort(tmp string) {
	_ = os.Remove(tmp)
}

// ensureDir creates dir (and parents) if needed, ignoring the case where it already exists.
func ensureDir(dir string) error {
	return os.MkdirAll(dir, 0o755)
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func dirExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}
